README.md
# sdc-ssrp


ChangeLog
2008-03-23 Borut Razem <borut.razem AT siol.net>

	* src/pic/gen.h: Q&D solution for the bug, found by Jim Paris
	  and reverted it, since it was already fixed by Raphael ;-)
	* doc/knownbugs.html: updated for RC2

2008-03-22 Raphael Neider <rneider AT web.de>

	* src/SDCCicode.c (getPtrType): avoid unnecessary extra assignments
	  for PIC16 targets

2008-03-22 Maarten Brock <sourceforge.brock AT dse.nl>

	* src/mcs51/gen.c (adjustArithmeticResult): fixed bug 1839299
	* support/regression/tests/bug1839277.c: added related testBug1839299

2008-03-21 Maarten Brock <sourceforge.brock AT dse.nl>

	* src/ds390/gen.c (opIsGptr, adjustArithmeticResult, genAddrOf, genAssign),
	* src/mcs51/gen.c (opIsGptr, adjustArithmeticResult, genAddrOf, genAssign):
	  fixed bug 1839277
	* src/mcs51/gen.c: throughout only output hex constants
	* src/SDCCicode.c (getPtrType, geniCodeCast): fixed code size regression
	* support/regression/tests/bug1839277.c: new, added

2008-03-21 Philipp Klaus Krause <pkk AT spth.de>

	* src/z80/ralloc.c (callFuncByName):
	  Used // instead of /**/ to fix compilation warning about nested comments.

2008-03-20 Maarten Brock <sourceforge.brock AT dse.nl>

	* src/SDCCsymt.c (comparePtrType): fixed bug 1921073
	* support/regression/tests/bug1921073.c: new, added

2008-03-17 Philipp Klaus Krause <pkk AT spth.de>

	* src/z80/SDCCpeeph.c.c (callFuncByName):
	  Commented function parameters.

2008-03-17 Philipp Klaus Krause <pkk AT spth.de>

	* src/z80/ralloc.c (packRegsForAccUse2):
	  disabled a broken accumulator packing optimization (fixed #1292721).
	* support/regression/tests/bug-1292721: Regression test for #1292721.

2008-03-16 Raphael Neider <rneider AT web.de>

	* src/pic/gen.c (SetIrp,genNearPointerGet,genNearPointerSet):
	  fixed code bloat regression using a variant of patch #1915618

2008-03-16 Borut Razem <borut.razem AT siol.net>

	* support/scripts/sdcc.nsi: added section debugging macros, added SDCC
	  icon, changed startup menu for release distribution, code cleaning,
	  create url files instead links in startup menu

2008-03-16 Maarten Brock <sourceforge.brock AT dse.nl>

	* src/SDCCcse.c (algebraicOpts): fixed bug 1579949
	* support/regression/tests/bug1579949.c: new, added

2008-03-14 Paul Stoffregen <paul AT pjrc.com>

	* device/lib/printf_fast.c: fixed bug 1255403

2008-03-14 Maarten Brock <sourceforge.brock AT dse.nl>

	* doc/sdccman.lyx: enhanced paragraph about bankswitching for mcs51
	* src/cdbFile.c (cdbWriteBasicSymbol): fixed bug 1909409 for locals

2008-03-14 Borut Razem <borut.razem AT siol.net>

	* as/doc/asmlnk.doc, as/doc/asxhtm.html: updated link-xx and aslink
	  command line options
	* as/link/hc08/lkmain.c: converted tabs to spaces so that the usage
	  is correctly displayed
	* as/link/z80/lkmain.c: usage line commented out

2008-03-13 Borut Razem <borut.razem AT siol.net>

	* doc/sdccman.lyx: corrected to be able to convert to PDF,
	  updated the list pf supported pic14 devices

2008-03-12 Maarten Brock <sourceforge.brock AT dse.nl>

	* doc/sdccman.lyx: added paragraph about bankswitching for mcs51

2008-03-12 Borut Razem <borut.razem AT siol.net>

	* as/doc/asmlnk.doc, as/doc/asxhtm.html: updated command line options,
	  fixed documentation request #1718191
	* as/hc80/asmain.c, as/mcs51/asmain.c: added option -c to usage
	* doc/sdccman.lyx: added description of --no-std-crt0 gbz80
	  command line option

2008-03-10 Maarten Brock <sourceforge.brock AT dse.nl>

	* src/SDCCast.c (isConformingBody): fixed bug 1505811, thanks Robert Larice
	* support/regression/tests/bug1505811.c: new, added

2008-03-09 Raphael Neider <rneider AT web.de>

	* device/include/pic16/pic18f2620.h,
	* device/include/pic16/pic18f4620.h,
	* device/lib/pic16/libdev/pic18f2620.c,
	* device/lib/pic16/libdev/pic18f4525.c,
	* device/lib/pic16/libdev/pic18f4620.c: merged 18f2525/2620/4525/4620
	  family to consistently use pic18f4620.[ch], fixes #1832562

2008-03-09 Borut Razem <borut.razem AT siol.net>

	* .version, sdcc.spec, doc/sdccman.lyx, doc/INSTALL.txt:
	  changed sdcc version to 2.8.0
	* knownbugs.html: regenerated
	* support/scripts/gen_known_bugs.pl: added number of open bugs

2008-03-09 Raphael Neider <rneider AT web.de>

	* src/pic16/gen.c (pic16_mov2w_volatile): NEW, read volatile operands,
	  (genOr): read volatile operand even if the result is known,
	  closes #1511838
	  (genOr,genXor): removed unused legacy code,
	  (genDummyRead): use pic16_mov2w_volatile()

2008-03-08 Borut Razem <borut.razem AT siol.net>

	* src/SDCCglue.c: fixed bug #1864582: multiple definition of char
	  cons w. --model-large
	* support/regression/staticinit.c: added regression test for bug
	  #1864582

2008-03-08 Maarten Brock <sourceforge.brock AT dse.nl>

	* src/ds390/gen.c (pushSide, genPcall),
	* src/hc08/gen.c (pushSide, genPcall): synchronized with mcs51
	* src/mcs51/gen.c: cosmetic changes
	* support/regression/fwk/include/testfwk.h: added macro reentrant
	* support/regression/tests/bug1908493.c: new, added

2008-03-08 Jesus Calvino-Fraga <jesusc AT ece.ubc.ca>

	* src/SDCCdebug.c:, as/link/lkaomf51.c: Fixed bug 1909409: Pdata in OMF file

2008-03-08 Maarten Brock <sourceforge.brock AT dse.nl>

	* src/mcs51/gen.c (pushSide, genPcall): fixed bug 1908493
	* src/SDCCmain.c (preProcess): removed -DSDCC_PARMS_IN_BANK1 because it is
	  already set in ds390/main.c and mcs51/main.c

2008-03-07 Raphael Neider <rneider AT web.de>

	* src/regression/init0.c: new test for initialized arrays of function
	  pointers
	* src/regression/Makefile: made a bit more flexible, added init0.c
	* src/pic/glue.c (emitIvals): hacky fix for initializing from function
	  pointers, closes #1427663

2008-03-05 Borut Razem <borut.razem AT siol.net>

	* dos/sdccman.lyx: docummented predefined macros SDCC_REVISION,
	  SDCC_PARMS_IN_BANK1, SDCC_FLOAT_REENT and SDCC_INT_LONG_REENT

2008-03-05 Jesus Calvino-Fraga <jesusc AT ece.ubc.ca>

	* support/librarian/sdcclib.c:Added feature request 1908061:

device/lib/pic16/libc/stdio/vfprintf.c
/*-----------------------------------------------------------------
    vfprintf.c - source file for reduced version of printf

    Modified for pic16 port, by Vangelis Rokas, 2005 (vrokas@otenet.gr)

    Written By - Sandeep Dutta . sandeep.dutta@usa.net (1999)

    This library is free software; you can redistribute it and/or modify it
    under the terms of the GNU Library General Public License as published by the
    Free Software Foundation; either version 2, or (at your option) any
    later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Library General Public License for more details.

    You should have received a copy of the GNU Library General Public License
    along with this program; if not, write to the Free Software
    Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

    In other words, you are welcome to use, share and improve this program.
    You are forbidden to forbid anyone else to use, share and improve
    what you give them.   Help stamp out software-hoarding!
-------------------------------------------------------------------------*/

/*
** $Id: vfprintf.c 5008 2008-02-13 21:59:32Z MaartenBrock $
*/

/* following formats are supported :-
   format     output type       argument-type
     %%        -                   -
     %u*       unsigned            *
     %b	       binary              int
     %lb       binary              long
     %hb       binary              char
     %d        decimal             int
     %ld       decimal             long
     %hd       decimal             char
     %[xX]     hexadecimal         int
     %l[xX]    hexadecimal         long
     %h[xX]    hexadecimal         char
     %o        octal               int
     %lo       octal               long
     %ho       octal               char
     %c        character           char
     %s        character           generic pointer
*/

#include <ctype.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>

#if _DEBUG
extern void io_long(unsigned long);
extern void io_str(char *);
extern void io_int(unsigned int);
#endif

unsigned int vfprintf(FILE *stream, char *fmt, va_list ap)
{
  unsigned char radix;
  unsigned char flong, fstr;
  unsigned char fchar, nosign;
  unsigned char upcase;
  unsigned int count=0;
  char *str, *ch;
  __data char *str1;
  long val;
//  static char buffer[16];
  char buffer[16];


#if _DEBUG
  io_str( "vfprintf: " );
  io_long( (unsigned long)stream );
  io_long( (unsigned long)fmt );
#endif

//    va_start(ap,fmt);
  ch = fmt;

  while( *ch ) {      //for (; *fmt ; fmt++ )
    if (*ch == '%') {
      flong = 0;
      fstr = 0;
      fchar = 0;
      nosign = 0;
      radix = 0;
      upcase = 0;
      ch++;

      if(*ch == '%') {
        __stream_putchar(stream, *ch);
        ++count;
        ++ch;
        continue;
      }

      if(*ch == 'u') {
        nosign = 1;
        ch++;
      }

      if(*ch == 'l') {
        flong = 1;
        ch++;
      } else if(*ch == 'h') {
        fchar = 1;
        ch++;
      }

      if(*ch == 's')fstr = 1;
      else if(*ch == 'd')radix = 10;
      else if(*ch == 'x'){ radix = 16; upcase = 0; }
      else if(*ch == 'X'){ radix = 16; upcase = 1; }
      else if(*ch == 'c')radix = 0;
      else if(*ch == 'o')radix = 8;
      else if(*ch == 'b')radix = 2;
      else {
        __stream_putchar(stream, *ch);
        ++count;
        ++ch;
        continue;
      }

      if(fstr) {
        str = va_arg(ap, char *);
        while(*str) { __stream_putchar(stream, *str); str++; count++; }
      } else {
        val = 0;
        if(flong) {
          val = va_arg(ap, long);
#if _DEBUG
          io_long(val);
#endif
        }
        else if(fchar) {
          val = (unsigned char)va_arg(ap, int); // FIXME: SDCC passes 1-byte char varargs as 2-byte ints...
	  if (!nosign) val = (char) val; // (FIXME cont'd) sign extend if required
#if _DEBUG
          io_long(val);
#endif
        }
        else {
          val = va_arg(ap, int);
#if _DEBUG
          io_long(val);
#endif
        }

        if(radix) {
          if(nosign)
            ultoa(val, buffer, radix);
          else
            ltoa (val, buffer, radix);

          str1 = buffer;
          while( (*str1) ) {
            radix = *str1;
            if(upcase)
              radix = toupper( radix );
            __stream_putchar(stream, (unsigned char)radix);
            count++;
            str1++;
          }
        } else {
          __stream_putchar(stream, (unsigned char)val);
          count++;
        }
      }
    } else {
      __stream_putchar(stream, *ch);
      count++;
    }

    ch++;
  }

  return (count);
}


device/lib/pic16/libc/stdio/printf_small.c
/*-----------------------------------------------------------------
    printf_small.c - source file for reduced version of printf

    Modified for pic16 port, by Vangelis Rokas, 2004 (vrokas@otenet.gr)
    
    Written By - Sandeep Dutta . sandeep.dutta@usa.net (1999)

    This library is free software; you can redistribute it and/or modify it
    under the terms of the GNU Library General Public License as published by the
    Free Software Foundation; either version 2, or (at your option) any
    later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Library General Public License for more details.

    You should have received a copy of the GNU Library General Public License
    along with this program; if not, write to the Free Software
    Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

    In other words, you are welcome to use, share and improve this program.
    You are forbidden to forbid anyone else to use, share and improve
    what you give them.   Help stamp out software-hoarding!
-------------------------------------------------------------------------*/

/*
** $Id: printf_small.c 4511 2006-12-10 17:07:41Z tecodev $
*/

/* This function uses function putchar() to dump a character
 * to standard output. putchar() is defined in libc18f.lib
 * as dummy function, which will be linked if no putchar()
 * function is provided by the user.
 * The user can write his own putchar() function and link it
 * with the source *BEFORE* the libc18f.lib library. This way
 * the linker will link the first function (i.e. the user's function) */

/* following formats are supported :-
   format     output type       argument-type
     %d        decimal             int
     %ld       decimal             long
     %hd       decimal             char
     %x        hexadecimal         int
     %lx       hexadecimal         long
     %hx       hexadecimal         char
     %o        octal               int
     %lo       octal               long
     %ho       octal               char
     %c        character           char
     %s        character           generic pointer
     %f        float               float
*/

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>

void printf_small(char *fmt, ...) __reentrant
{
  char *ch;
  char radix;
  char flong;
  char fstr;
  char fchar;
  char ffloat;
  float flt;
  char *str;
  __data char *str1;
  long val;
  static char buffer[16];
  va_list ap ;

    ch = fmt;
    va_start(ap,fmt);

    while( *ch ) {			//for (; *fmt ; fmt++ )
        if (*ch == '%') {
            flong = fstr = fchar = ffloat = 0;
            radix = 0;
            ch++;

            if(*ch == 'l') {
              flong = 1;
              ch++;
            } else
            if(*ch == 'h') {
              fchar = 1;
              ch++;
            }
            
            if(*ch == 's')fstr = 1;
            else if(*ch == 'f')ffloat = 1;
            else if(*ch == 'd')radix = 10;
            else if(*ch == 'x')radix = 16;
            else if(*ch == 'c')radix = 0;
            else if(*ch == 'o')radix = 8;
            
            if(fstr) {
                str = va_arg(ap, char *);
                while (*str) putchar(*str++);
            } else
            if(ffloat) {
                flt = va_arg(ap, float);
                x_ftoa(flt, buffer, 32, 6);
                str1 = buffer;
                while( *str1 )str1++; str1--;
                while( *str1 == '0' )str1--; str1++;
                *str1 = 0; str1 = buffer;
                while( *str1 )putchar(*str1++);
            } else {
              if(flong)val = va_arg(ap, long);
              else
              if(fchar)val = (char)va_arg(ap, int); // FIXME: SDCC casts char arguments into ints
              else {
                  val = va_arg(ap, int);
              }
            
              if(radix) {
                ltoa (val, buffer, radix);

                str1 = buffer;
                while ( *str1 ) {
                  putchar ( *str1++ );
                }
              }	else
                putchar((char)val);
            }
          } else
            putchar(*ch);

        ch++;
    }
}


device/lib/pic16/libc/stdio/printf_tiny.c
/*-----------------------------------------------------------------
    printf_tiny.c - source file for reduced version of printf

    Modified for pic16 port, by Vangelis Rokas, 2004 (vrokas@otenet.gr)
    
    Written By - Sandeep Dutta . sandeep.dutta@usa.net (1999)

    This library is free software; you can redistribute it and/or modify it
    under the terms of the GNU Library General Public License as published by the
    Free Software Foundation; either version 2, or (at your option) any
    later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Library General Public License for more details.

    You should have received a copy of the GNU Library General Public License
    along with this program; if not, write to the Free Software
    Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

    In other words, you are welcome to use, share and improve this program.
    You are forbidden to forbid anyone else to use, share and improve
    what you give them.   Help stamp out software-hoarding!
-------------------------------------------------------------------------*/

/*
** $Id: printf_tiny.c 4511 2006-12-10 17:07:41Z tecodev $
*/

/* This function uses function putchar() to dump a character
 * to standard output. putchar() is defined in libc18f.lib
 * as dummy function, which will be linked if no putchar()
 * function is provided by the user.
 * The user can write his own putchar() function and link it
 * with the source *BEFORE* the libc18f.lib library. This way
 * the linker will link the first function (i.e. the user's function) */

/* following formats are supported :-
   format     output type       argument-type
     %u*       unsigned            *

     %b	       binary
     %d        decimal             int
     %ld       decimal             long
     %hd       decimal             char
     %x        hexadecimal         int
     %lxX      hexadecimal         long
     %hxX      hexadecimal         char
     %o        octal               int
     %lo       octal               long
     %ho       octal               char
     %c        character           char
     %s        character           generic pointer
*/

#include <ctype.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>

#if 0
#define	DPUT(c)	putchar( c )
#else
#define DPUT(c)
#endif

#define ISLONG		(flong)
#define ISSTR		(fstr)
#define ISCHAR		(fchar)
#define HAVESIGN	(nosign)


#if 1
extern void io_long(long);
extern void io_int(long);
extern void io_char(char);
#endif

void printf_tiny(char *fmt, ...) 
{
  char radix;
  char flong, fstr;
  char fchar, nosign;
  char upcase;

  char *str, *ch;
  __data char *str1;
  long val;
//  static char buffer[16];
  char buffer[16];
  va_list ap ;

    va_start(ap,fmt);
    ch = fmt;
    
    while( *ch ) {			//for (; *fmt ; fmt++ )
        if (*ch == '%') {
            ISLONG = 0;
            ISSTR = 0;
            ISCHAR = 0;
            HAVESIGN = 0;
            radix = 0;
            upcase = 0;
            ch++;

            if(*ch == 'u') {
              HAVESIGN = 1;
              ch++;
            }
            
            if(*ch == 'l') {
              ISLONG = 1;
              ch++;
            } else
            if(*ch == 'h') {
              ISCHAR = 1;
              ch++;
            }
            
            if(*ch == 's')ISSTR = 1;
            else if(*ch == 'd')radix = 10;
            else if(*ch == 'x'){ radix = 16; upcase = 0; }
            else if(*ch == 'X'){ radix = 16; upcase = 1; }
            else if(*ch == 'c')radix = 0;
            else if(*ch == 'o')radix = 8;
            else if(*ch == 'b')radix = 2;
            
            if(ISSTR) {
                str = va_arg(ap, char *);
                while(*str) { putchar(*str);str++;}
            } else {
              if(ISLONG)val = va_arg(ap, long);
              else
              if(ISCHAR) {
		  val = (unsigned char)va_arg(ap, int); // FIXME: SDCC casts char arguments into ints
		  if (!HAVESIGN) val = (char)val; // FIXME cont'd: sign-extend if required
	      } else {
                  val = va_arg(ap, int);
              }

              if(radix) {
                if(HAVESIGN)
                  ultoa(val, buffer, radix);
                else
                  ltoa (val, buffer, radix);

                str1 = buffer;
                while( (*str1) ) {
                  radix = *str1;
                  if(upcase)
                    radix = toupper( radix );
                  putchar ( radix );
                  str1++;
                }
              }	else
                putchar((char)val);
            }
          } else
            putchar(*ch);

        ch++;
    }
}
